GH-43577: [Java] getBuffers method needs correction on clear flag usage

vibhatha opened the pull request:
### Rationale for this change

`getBuffers` method provides the capability to clear the buffers in the vector, this has not been properly tested while clear flag is not properly used in the implementation across various types of vectors. 

### What changes are included in this PR?

Updating the vector `getBuffers` method to use `clear` flag as expected and adding corresponding test cases. 

### Are these changes tested?

Yes, via existing test cases and new test cases. 

### Are there any user-facing changes?

Yes
* GitHub Issue: #43577

lidavidm commented:
I think this should be considered a breaking change

lidavidm commented:
Also, are we certain the original behavior wasn't intentional? How is clear currently used?

vibhatha commented:
> I think this should be considered a breaking change

Yes it is. Sorry I didn't make it.

vibhatha commented:
> Also, are we certain the original behavior wasn't intentional? How is clear currently used?

This came up in the `LargeListVIewVector` review. Up to that point, I also didn't think about it that much, I was actually using the existing logic from other vectors. Clear is just used to determine whether or not to clear the vector but not the `FieldVector` assoicated with that vector.

vibhatha commented:
If we allow this, we need to document and remind the user that when clear is true, there is a responsibility in clearing the retrieved buffers.

lidavidm commented:
> > Also, are we certain the original behavior wasn't intentional? How is clear currently used?
> 
> This came up in the `LargeListVIewVector` review. Up to that point, I also didn't think about it that much, I was actually using the existing logic from other vectors. Clear is just used to determine whether or not to clear the vector but not the `FieldVector` assoicated with that vector.

Right, I pointed it out since it didn't make immediate sense. But I'm asking if the codebase gives a reason why clear works this way. It might not be obvious.

vibhatha commented:
Let me investigate this a little more.

vibhatha commented:
Looking into this, I think, clearing the buffers of the child vector might release memory that is still in use elsewhere. And this might lead to unexpected behavior or errors as we are explicitly enforcing a buffer clear on the extracted ones. Also, I think the clear was only meant to be used to clear the vector's buffer not its child buffers. But I feel like this is not precisely documented, as it only says "Return the underlying buffers associated with this vector".

vibhatha commented:
Reconsidering my position, I would like to propose a documentation update and reverting the change made for `LargeListViewVector` in addressing reviews in the previous PR.

lidavidm commented:
Sounds good to me. Thanks for looking!

vibhatha commented on java/vector/src/test/java/org/apache/arrow/vector/TestVectorReset.java:
```diff
@@ -136,6 +140,13 @@ public void testListTypeReset() {
       fixedList.setValueCount(1);
       resetVectorAndVerify(fixedList, fixedList.getBuffers(false));
+
+      // ListViewVector
```
@lidavidm I didn't add the `LargeList` and `LargeListView` as `LargeList` wasn't there in the first place.

lidavidm commented on java/vector/src/main/java/org/apache/arrow/vector/complex/AbstractStructVector.java:
```diff
@@ -382,12 +382,23 @@ public VectorWithOrdinal getChildVectorWithOrdinal(String name) {
 
     for (final ValueVector vector : vectors.values()) {
-      for (final ArrowBuf buf : vector.getBuffers(false)) {
+      for (final ArrowBuf buf : vector.getBuffers(clear)) {
```
We _don't_ want to do this now, right?

vibhatha replied:
yeah I have missed this.

vibhatha replied:
Haven't committed sorry.

vibhatha commented:
@lidavidm I updated the PR with the doc change.
